### PR TITLE
Code Review

### PR DESCRIPTION
**Pull request by willmot**
Master PR for all the various fixes that come out of code review.
- [x] Code Formatting
- [x] Security
- [x] Translations
- ~~Admin UI~~ Handled in #9

### CODE REVIEW

File: includes/render.php [hunk 1]
```diff
@@ -24,30 +25,33 @@ public function process( $form_id, $echo = false ) {
...
-					$show_powered_by = Mad_Mimi_Settings_Controls::get_option( 'display_powered_by' ) ? true : false;
+					<?php $show_powered_by = Mad_Mimi_Settings_Controls::get_option( 'display_powered_by' ) ? true : false;
 
 					if ( $show_powered_by ) : ?>
-					<p>
-						<a href="http://madmimi.com" target="_blank"><?php _e( 'Powered by Mad Mimi', 'mimi' ); ?></a>
-					</p>
+
+						<p>
+							<a href="http://madmimi.com" target="_blank"><?php _e( 'Powered by Mad Mimi', 'mimi' ); ?></a>
```
Comment (mattheu) [inline includes/render.php]: Should probably be using escaped translation functions eg `esc_html_e`

File: includes/render.php [hunk 2]
```diff
@@ -56,54 +60,71 @@ class Mad_Mimi_Form_Fields {
...
 		self::$cycle = absint( $cycle );
 
 		call_user_func( array( __CLASS__, $field->type ), $field );
+
 	}
 
 	public function get_form_id( $field_name ) {
+
 		// since HTML ID's can't exist in the same exact spelling more than once... make it special.
 		return esc_attr( sprintf( 'form_%s_%s', self::$cycle, $field_name ) );
```
Comment (mattheu) [inline includes/render.php]: I know you've not changed this line... but this escape call is redundant really - it should be done when its output.

Comment (mattheu) [inline includes/render.php]: which it isn't...

File: includes/render.php [hunk 3]
```diff
@@ -56,54 +60,71 @@ class Mad_Mimi_Form_Fields {
...
 			<?php echo esc_html( $args->display ); ?>
+
 			<?php if ( $args->required && apply_filters( 'mimi_required_field_indicator', true, $args ) ) : ?>
-			<span class="required">*</span>
+				<span class="required">*</span>
 			<?php endif; ?>
+
 		</label>
+
 		<input type="text" name="<?php echo esc_attr( $args->name ); ?>" id="<?php echo self::get_form_id( $args->name ); ?>" class="<?php echo esc_attr( join( ' ', $field_classes ) ); ?>" />
```
Comment (mattheu) [inline includes/render.php]: `get_form_id` should be escaped

File: includes/render.php [hunk 4]
```diff
@@ -56,54 +60,71 @@ class Mad_Mimi_Form_Fields {
...
-		if ( $args->required )
+		if ( $args->required ) {
 			$field_classes[] = 'mimi-required';
+		}
+
+		$field_classes = (array) apply_filters( 'mimi_required_field_class', $field_classes, $args ); ?>
 
-		$field_classes = (array) apply_filters( 'mimi_required_field_class', $field_classes, $args );
-		?>
 		<label for="<?php echo self::get_form_id( $args->name ) . esc_attr( $args->value ); ?>">
```
Comment (mattheu) [inline includes/render.php]: `get_form_id` should be escaped

File: includes/render.php (hunk 3 above)
Comment (willmot) [inline includes/render.php]: 👍

File: includes/settings.php [hunk 5]
```diff
@@ -279,136 +315,153 @@ public function display_settings_page() {
...
+					<?php endif; ?>
+
 					</tbody>
 				</table>
 
 				<br />
+
 				<p class="description">
-					<?php _e( 'Not seeing your form?', 'mimi' ); ?> <a href="<?php echo add_query_arg( 'action', 'refresh' ); ?>" class="button"><?php _e( 'Refresh Forms', 'mimi' ); ?></a>
+					<?php _e( 'Not seeing your form?', 'mimi' ); ?> <a href="<?php echo esc_url( add_query_arg( 'action', 'refresh' ) ); ?>" class="button"><?php _e( 'Refresh Forms', 'mimi' ); ?></a>
```
Comment (mattheu) [inline includes/settings.php]: Should be using escaped translation functions here

File: includes/settings.php [hunk 6]
```diff
@@ -279,136 +315,153 @@ public function display_settings_page() {
...
 
-				<h3><?php _e( 'Debug', 'mimi' ); ?></h3>
-				<p>
-					<a href="<?php echo add_query_arg( 'action', 'debug-reset' ); ?>" class="button-secondary"><?php _e( 'Erase All Data', 'mimi' ); ?></a>
-					<a href="<?php echo add_query_arg( 'action', 'debug-reset-transients' ); ?>" class="button-secondary"><?php _e( 'Erase Transients', 'mimi' ); ?></a>
-				</p>
+					<h3><?php _e( 'Debug', 'mimi' ); ?></h3>
+					<p>
+						<a href="<?php echo esc_url( add_query_arg( 'action', 'debug-reset' ) ); ?>" class="button-secondary"><?php _e( 'Erase All Data', 'mimi' ); ?></a>
+						<a href="<?php echo esc_url( add_query_arg( 'action', 'debug-reset-transients' ) ); ?>" class="button-secondary"><?php _e( 'Erase Transients', 'mimi' ); ?></a>
```
Comment (mattheu) [inline includes/settings.php]: Should be using escaped translation functions here

File: includes/widget.php [hunk 7]
```diff
@@ -48,46 +53,52 @@ function update( $new_instance, $old_instance ) {
...
-		$selected_form = absint( $instance['form'] );
-		$text          = esc_textarea( $instance['text'] );
+		$forms = Mad_Mimi_Dispatcher::get_forms(); ?>
 
-		$forms = Mad_Mimi_Dispatcher::get_forms();
-		?>
 		<p>
-			<label for="<?php echo $this->get_field_id( 'title' ); ?>"><?php _e( 'Title:' ); ?></label>
-			<input class="widefat" id="<?php echo $this->get_field_id( 'title' ); ?>" name="<?php echo $this->get_field_name( 'title' ); ?>" type="text" value="<?php echo $title; ?>" />
+			<label for="<?php echo esc_attr( $this->get_field_id( 'title' ) ); ?>"><?php _e( 'Title:', 'mimi' ); ?></label>
```
Comment (mattheu) [inline includes/widget.php]: Should be using escaped translation functions here

File: includes/widget.php [hunk 8]
```diff
@@ -48,46 +53,52 @@ function update( $new_instance, $old_instance ) {
...
-						<option value="<?php echo esc_attr( $f->id ); ?>" <?php selected( $selected_form, $f->id ); ?>><?php echo esc_html( $f->name ); ?></option>
+						<option value="<?php echo esc_attr( $f->id ); ?>" <?php selected( $instance['form'], $f->id ); ?>><?php echo esc_html( $f->name ); ?></option>
 					<?php endforeach; ?>
+
 				</select>
 
 			<?php else : ?>
 
-			<span><?php printf( __( 'Please set up your Mad Mimi account in the <a href="%s">settings page</a>.', 'mimi' ), admin_url( 'options-general.php?page=mad-mimi-settings' ) ); ?>
+			<span><?php printf( __( 'Please set up your Mad Mimi account in the <a href="%s">settings page</a>.', 'mimi' ), esc_url( admin_url( 'options-general.php?page=mad-mimi-settings' ) ) ); ?>
```
Comment (mattheu) [inline includes/widget.php]: Its situations like this that @rmccue made his `whitelist_html` function. You could use wp_kses here if you didn't want to include his code. Its pretty quick if you restrict to only a short string with a small selection of allowed tags.

Comment (willmot) [inline includes/widget.php]: I think I'll just go with `wp_kses` 👍

File: js/mimi.js [hunk 9]
```diff
@@ -82,27 +89,33 @@
...
+
 			$.each( type, function( index, value ) {
 				_class.push( 'mimi-' + value );
 			} );
+
 		} else {
 			_class.push( 'mimi-' + type.toString() );
 		}
 
 		return $( '<p/>', { class: _class.join( ' ' ) } ).text( message );
```
Comment (mattheu) [inline js/mimi.js:117]: Nice - this is the best way to create elements in JS and avoid XSS vulnerabilities

**Comment by mattheu** [comment]
This is all looking pretty good.

Pointed out a few places where you could be using escaped translation functions. I didn't actually flag them all - but look out for for `_e` and `echo __`

**Comment by willmot** [comment]
Thanks Matt I think I've got all the translation functions properly escaped now 👍

**Comment by willmot** [comment]
Choo Choo 🚄

File: includes/settings.php [hunk 10]
```diff
@@ -21,82 +21,108 @@ public function __construct() {
...
+				case 'edit_form' :
+
+					if ( ! isset( $_GET['form_id'] ) ) {
 						return;
+					}
 
 					$tokenized_url = add_query_arg( 'redirect', sprintf( '/signups/%d/edit', absint( $_GET['form_id'] ) ), Mad_Mimi_Dispatcher::user_sign_in() );
 
+					// Not wp_safe_redirect as it's an external site
 					wp_redirect( $tokenized_url );
```
Comment (joehoyle) [inline includes/settings.php:107]: Can this use `wp_safe_redirect`

Comment (willmot) [inline includes/settings.php:107]: @joehoyle it's redirecting to an external site, I guess we'd need to add that domain to the safe redirect list?

Comment (joehoyle) [inline includes/settings.php:107]: Ah in that case, I'd leave it as is

**Pull request merged**